File: frontend/src/service/table/Cancelation.ts
import { z } from 'zod'

import { LineItemSchema } from './Order'

export const CancelationSchema = z.object({
  id: z.uuid(),
  userId: z.number().int().min(1),
  tableId: z.number().int().min(1),
  variants: LineItemSchema.array().min(1),
  totalCancelationCents: z.number().int().min(0),
  comment: z.string().max(100),
  canceledAt: z.string().refine((date) => !isNaN(Date.parse(date)), {
    message: 'Invalid date format',
  }),
})
export type Cancelation = z.infer<typeof CancelationSchema>

export const CancelVariantsSchema = z.object({
  tableId: z.number().int().min(1),
  variants: LineItemSchema.array().min(1),
  comment: z.string().max(100),
})
export type CancelVariants = z.infer<typeof CancelVariantsSchema>

File: backend/domain/table/cancelation.go
package table

import (
	"time"

	z "github.com/Oudwins/zog"
)

type Cancelation struct {
	ID                    string     `json:"id"`
	UserID                int        `json:"userId"`
	TableID               int        `json:"tableId"`
	Variants              []LineItem `json:"variants"`
	TotalCancelationCents int        `json:"totalCancelationCents"`
	Comment               string     `json:"comment"`
	CanceledAt            time.Time  `json:"canceledAt"`
}

var cancelationSchema = z.Struct(z.Shape{
	"ID":                    z.String().UUID().Required(),
	"UserID":                z.Int().GTE(1).Required(),
	"TableID":               z.Int().GTE(1).Required(),
	"Variants":              z.Slice(lineItemSchema).Min(1).Required(),
	"TotalCancelationCents": z.Int().GTE(0).Required(),
	"Comment":               z.String().Max(100),
	"CanceledAt":            z.Time().Required(),
})
